### PR TITLE
Shadow control for zones, models and shapes

### PR DESCRIPTION
**Pull request by NissimHadar**
# General
This PR enables content creator control of casting shadows.  
1.  Shadows may be turned on/off per zone
2.  Shadows may be turned on/off per entity/shape.
# Requirements
1. Master shadow flag defaults to on (Menu->Developer->Graphics...) DONE
2. Add a shadow cast flag to the keylight component of the zone, defaulting to off DONE
3. Do not run shadow task when shadows are not active.
4. Add a shadow cast flag to Entity, dafaulting to on.  Entities will cast shadows iff the master & active zone & entity flags are all active DONE
5. Avatars will always cast shadows (provided that the master & the active zone flags are on DONE
6. Shadows/Ambient Occlusion global control - use menu checkboxes instead of qml screen.
# Testing
Tests are defined in: https://highfidelity.testrail.net/index.php?/suites/view/755&group_by=cases:title&group_order=asc

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7723/app_12558_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7723/HighFidelity-Beta-PR12558-1eb0dde3d23a987fd7c727d7eaadc945c683f620.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7723/HighFidelity-Beta-PR12558-1eb0dde3d23a987fd7c727d7eaadc945c683f620.dmg

File: libraries/entities/src/EntityPropertyFlags.h [hunk 1]
```diff
@@ -254,10 +260,6 @@ enum EntityPropertyList {
     // Aliases/Piggyback properties for Zones. These properties intentionally reuse the enum values for
     // other properties which will never overlap with each other. We do this so that we don't have to expand
     // the size of the properties bitflags mask
-    PROP_KEYLIGHT_COLOR = PROP_COLOR,
```
Comment (samcake) [inline libraries/entities/src/EntityPropertyFlags.h]: Can't you keep these reused property Enums here and find another one to reuse for "Prop_KEYLIGHT_CAST_SHADOW" ?

**Comment by tapalisa** [comment]
I agree with Liv. We went through a lot of work to reorder the properties so that they are roughly in an order of usefulness.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7796/app_12558_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7796/HighFidelity-Beta-PR12558-d72f46ec3e413a457167831071c59f93a43a57b3.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7796/HighFidelity-Beta-PR12558-d72f46ec3e413a457167831071c59f93a43a57b3.dmg

File: libraries/entities-renderer/src/RenderableEntityItem.cpp [hunk 2]
```diff
@@ -164,7 +164,12 @@ ItemKey EntityRenderer::getKey() {
         return ItemKey::Builder::transparentShape().withTypeMeta().withTagBits(render::ItemKey::TAG_BITS_0 | render::ItemKey::TAG_BITS_1);
     }
 
-    return ItemKey::Builder::opaqueShape().withTypeMeta().withTagBits(render::ItemKey::TAG_BITS_0 | render::ItemKey::TAG_BITS_1);
+    // This allows shapes to cast shadows
+    if (_canCastShadow) {
+        return ItemKey::Builder::opaqueShape().withTypeMeta().withTagBits(render::ItemKey::TAG_BITS_0 | render::ItemKey::TAG_BITS_1).withShadowCaster();
```
Comment (SamGondelman) [inline libraries/entities-renderer/src/RenderableEntityItem.cpp]: Should we use another tag bit for this?  This is basically equivalent to "render from the shadow camera."

Comment (NissimHadar) [reply libraries/entities-renderer/src/RenderableEntityItem.cpp]: I don't understand those tag bits.  What does each one do (TAG_BIT_0 etc)

Comment (samcake) [reply libraries/entities-renderer/src/RenderableEntityItem.cpp]: the tags are just custom flags to filter the render items
We use it from the View  side to say:
Main view render anything that has Tag0
2ndary Camera render Anything that has Tag1

Comment (samcake) [reply libraries/entities-renderer/src/RenderableEntityItem.cpp]: THe shadow bit is a built in feature of render items when represented inside a view, so we have a built in flag for it.
It s equivalent to say "Tag3" is for shadow 
Except that there is a hardcoded meaning to that flag

**Comment by Qliemillar** [comment]
Needs a milestone before we can test, please!

**Comment by samcake** [comment]
@Qliemillar Please test now, thanks

**Comment by qlimnavarro** [comment]
Currently Testing.

**Comment by qlimnavarro** [comment]
Fail @conklin94122 
When Shadows is checked in the Developer menu and local Zone has shadows enabled, the Avatar does not produce shadows. In addition to this selecting ambient occlusion crashes the interface. Video and Logs attached:
[Files.zip](https://github.com/highfidelity/hifi/files/1794384/Files.zip)
Does not occur in Master.

**Comment by NissimHadar** [comment]
I am testing.

**Comment by marko8904** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7886/app_12558_PR.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7887/app_12558_PR.apk

**Comment by marko8904** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7890/app_12558_PR.apk

**Comment by NissimHadar** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7892/app_12558_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7892/HighFidelity-Beta-PR12558-fda32e4aa4bf267d6781617d4827781f60e92ecf.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7892/HighFidelity-Beta-PR12558-fda32e4aa4bf267d6781617d4827781f60e92ecf.exe

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by qlimnavarro** [comment]
Pass @conklin94122

**Comment by qlimnavarro** [comment]
Pass @conklin94122

**Comment by samcake** [comment]
Let's do it, thanks  all!

**Pull request merged**